### PR TITLE
FIX a single quote in a worksheet name issue

### LINKED ISSUE #605: Crash when opening a Worksheet named with a simple quote
**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

e.g. 0.90.0

**What is the current behavior?**
ClosedXml crash when opening a workbook that contains a worksheet whose name contains a simple quote and, if this sheet has a named range.
In the given file, if I remove the apostrophe in the name, it doesn't crash anymore.

## Reproducibility
**Code to reproduce problem:**
```c#
            var path = "\Error2.xlsx";
            var wb = new XLWorkbook(path);
```
[Error2.xlsx](https://github.com/ClosedXML/ClosedXML/files/1506634/Error2.xlsx)
- [x] I attached a sample spreadsheet.

### PR DESCRIPTION
**Pull request by Pankraty**
This PR fixes #605 

When saving a workbook with sheets which names contain a single quote sign (aka apostrophe) Excel escapes it with another quote and puts the sheet name into single quotes too. So the worksheet L'E from the original issue is represented as 'L''E' in the inner XML. Apparently, to avoid unescaping issues Excel prohibits setting worksheet names either starting or ending with a single quote.

ClosedXML, however, does not do the same. It allows using a single quote in a worksheet name, it can save and open books with apostrophes in the middle of worksheet name, but it fails at opening such books if they were saved by Excel, as all single quotes are being washed in the ParseReference method.

This PR 

1. prohibits using a single quote at the beginning or end of a worksheet name,
2. establishes proper escaping and unescaping quotes on loading and saving workbook,
3. adds tests covering these features.

I did not cover with an automatic test whether the references are escaped correctly in a resulted file, although I have checked this manually, and Excel is satisfied with the result too (initially, it complained that file has to be recovered).

- [ ] C# Code Review: @csreviewer
- [ ] Test Automation Review: @csreviewer

### CODE REVIEW

File: ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs [hunk 1]
```diff
@@ -234,5 +234,83 @@ public void CanCopySheetsWithAllAnchorTypes()
...
+
+            Assert.DoesNotThrow(saveAndOpenWorkbook);
+            Assert.AreEqual(title, savedTitle);
+        }
+
+        /// <summary>
+        /// Excel escapes symbol ' in worksheet title so we have to process this correctly.
+        /// </summary>
+        [Test]
+        public void CanOpenWorksheetWithEscapedApostrophe()
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs]: I think this test belongs with the others in `.\ClosedXML_Tests\Excel\Loading\LoadingTests.cs`. Do you agree?

Comment (Pankraty) [reply ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs]: Yes, I agree. Will move it.

File: ClosedXML/Excel/XLWorkbook_Load.cs [hunk 2]
```diff
@@ -1324,11 +1324,26 @@ private static void ParseReference(string item, out string sheetName, out string
...
         }
 
+        private static string UnescapeName(string worksheetName)
+        {
+            return worksheetName
+                .Trim('\'')
+                .Replace("''", "'");
+        }
+
+        private static string EscapeName(string worksheetName)
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook_Load.cs]: I think you should consolidate this function into `StringExtensions.WrapSheetNameInQuotesIfRequired` and rename it to `EscapeSheetName`.

File: ClosedXML/Excel/XLWorkbook_Save.cs [hunk 3]
```diff
@@ -731,7 +731,7 @@ orderby worksheet.Position
                     var definedNameText = worksheet.PageSetup.PrintAreas.Aggregate(String.Empty,
                         (current, printArea) =>
                             current +
-                            ("'" + worksheetName + "'!" +
+                            (EscapeName(worksheetName) + "!" +
```
Comment (igitur) [inline ClosedXML/Excel/XLWorkbook_Save.cs]: The single quotes are necessary when the sheet name has a space too.

Comment (igitur) [reply ClosedXML/Excel/XLWorkbook_Save.cs]: That's why some tests are failing.

Comment (Pankraty) [reply ClosedXML/Excel/XLWorkbook_Save.cs]: I see. And that is why some of the existing tests fail.

File: ClosedXML/Excel/XLWorkbook_Load.cs (hunk 2 above)
Comment (igitur) [reply ClosedXML/Excel/XLWorkbook_Load.cs]: If you do that, it should fix the failing tests.

**Comment by igitur** [comment]
Also, you seem like you're comfortable with git. When all tests and I approve the changes, would you mind squashing all the commits into one and rebasing? Going forward, I'm attempting to make the commit history cleaner.

**Comment by Pankraty** [comment]
Well, not that I am "comfortable" with Git, but I do my best to learn it (all my previous projects used SVN). I agree with about having a cleaner history.

**Comment by Pankraty** [comment]
With this commit all tests except 2 are green. As I see they both compare the result to the saved files (AutoFilter.xlsx and AdjustToContentsWithAutoFilter.xlsx). Sheet names in those files are escaped in quotes even though they don't have spaces or quotes in. Saving them by Excel removes the escaping characters. Shall I update these two files too?

**Comment by vbjay** [comment]
When you merge, you can tell it to squash.
https://github.com/blog/2141-squash-your-commits

**Comment by igitur** [comment]
@vbjay But I think (not sure) that if I do it, the commit line will be under my name instead of the actual collaborator, which is unfair. Am I wrong?

**Comment by vbjay** [comment]
No.  

image

image

**Comment by igitur** [comment]
What happens if commits from different contributors are squashed? Which one gets the squash commit?

**Comment by vbjay** [comment]
Good question.

**Comment by igitur** [comment]
Please rebase again (since `develop` branch has evolved).

**Comment by Pankraty** [comment]
> Please rebase again (since develop branch has evolved).

Done

**Pull request merged**